File: java/src/main/java/com/company/automation/support/RepoPaths.java
package com.company.automation.support;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Locates repository-level directories so the module works the same whether Maven runs from the
 * repo root or from inside {@code java/}, and whether an IDE sets the working directory to either.
 *
 * <p>Everything is resolved by walking up from the working directory looking for a marker, rather
 * than by hardcoding {@code ../}. A relative path that only works from one launch directory is the
 * most common reason a suite runs in the IDE but not in CI.
 */
public final class RepoPaths {

  private static final String MARKER = "shared/environments";
  private static final int MAX_LEVELS_UP = 6;

  private RepoPaths() {}

  /**
   * The repository root: the nearest ancestor directory that contains {@code shared/environments}.
   */
  public static Path repoRoot() {
    Path current = Paths.get("").toAbsolutePath().normalize();

    for (int level = 0; level <= MAX_LEVELS_UP; level++) {
      if (current == null) {
        break;
      }
      Path candidate = current.resolve(MARKER);
      if (Files.isDirectory(candidate)) {
        return current;
      }
      current = current.getParent();
    }

    throw new IllegalStateException(
        "Could not locate the repository root. Looked for a '"
            + MARKER
            + "' directory in the working directory and up to "
            + MAX_LEVELS_UP
            + " parents, starting at "
            + Paths.get("").toAbsolutePath().normalize()
            + ". Run Maven from the repository root or from the java/ directory.");
  }

  /** Directory holding the non-secret environment descriptors. */
  public static Path environments() {
    return repoRoot().resolve("shared").resolve("environments");
  }

  /** Directory holding the JSON Schema contracts. */
  public static Path contracts() {
    return repoRoot().resolve("shared").resolve("contracts");
  }

  /** Directory for this run's reports and failure artifacts. */
  public static Path reports() {
    return repoRoot().resolve("java").resolve("target");
  }
}

File: csharp/src/AutomationFramework.Core/Support/RepoPaths.cs
namespace AutomationFramework.Core.Support;

/// <summary>
/// Locates repository-level directories so the module behaves the same whether the test host runs
/// from the repository root, from <c>csharp/</c>, or from a build output folder several levels down.
/// </summary>
/// <remarks>
/// Everything is resolved by walking up from the current directory looking for a marker, rather than
/// by hardcoding <c>../../..</c>. A relative path that only works from one launch directory is the
/// most common reason a suite runs in the IDE but not from the command line or in CI.
/// </remarks>
public static class RepoPaths
{
    private const string Marker = "shared/environments";
    private const int MaxLevelsUp = 10;

    /// <summary>The repository root: the nearest ancestor containing <c>shared/environments</c>.</summary>
    public static string RepoRoot()
    {
        string? current = Directory.GetCurrentDirectory();

        for (int level = 0; level <= MaxLevelsUp; level++)
        {
            if (current is null)
            {
                break;
            }

            string candidate = Path.Combine(current, "shared", "environments");
            if (Directory.Exists(candidate))
            {
                return current;
            }

            current = Directory.GetParent(current)?.FullName;
        }

        throw new InvalidOperationException(
            $"Could not locate the repository root. Looked for a '{Marker}' directory in "
            + $"'{Directory.GetCurrentDirectory()}' and up to {MaxLevelsUp} parents. "
            + "Run dotnet from the repository root or from the csharp/ directory.");
    }

    /// <summary>Directory holding the non-secret environment descriptors.</summary>
    public static string Environments() => Path.Combine(RepoRoot(), "shared", "environments");

    /// <summary>Directory holding the JSON Schema contracts.</summary>
    public static string Contracts() => Path.Combine(RepoRoot(), "shared", "contracts");

    /// <summary>Directory for this run's reports and failure artifacts.</summary>
    public static string Reports() => Path.Combine(RepoRoot(), "csharp", "TestResults");
}
